Spigot-API-Patches/0136-Persistent-and-Temporary-Meta-API.patch

```diff
@@ -0,0 +1,676 @@
++        return getFloat(key, null);
++    }
++    public Float getFloat(@Nonnull K key, Float def) {
++        Number number = (Number) get(key);
++        return number != null ? number.floatValue() : def;
++    }
++    public Short getShort(@Nonnull K key) {
++        return getShort(key, null);
++    }
++    public Short getShort(@Nonnull K key, Short def) {
++        Number number = (Number) get(key);
++        return number != null ? number.shortValue() : def;
++    }
++    public Boolean getBoolean(@Nonnull K key) {
++        return getBoolean(key, null);
++    }
++    public Boolean getBoolean(@Nonnull K key, Boolean def) {
++        Number number = (Number) get(key);
++        return number != null ? number.intValue() != 0 : def;
++    }
++    public <T> T getValue(@Nonnull K key) {
++        //noinspection unchecked
++        return (T) get(key);
++    }
++    public <T> T getValue(@Nonnull K key, T def) {
++        Object value = get(key);
++        //noinspection unchecked
++        return value != null ? (T) value : def;
++    }
++
++    public Object get(K key) {
++        return get(key.key());
++    }
++
++    public Object put(MetaKey key, Object value) {
++        return put(key.key(), value);
++    }
++
++    @Override
++    public Object put(String key, Object value) {
++        return super.put(key, value);
++    }
++
++    public boolean containsKey(K key) {
++        return containsKey(key.key());
++    }
++}
+diff --git a/src/main/java/com/destroystokyo/paper/meta/PersistentMetaContainer.java b/src/main/java/com/destroystokyo/paper/meta/PersistentMetaContainer.java
+new file mode 100644
+index 000000000..f79db737f
+--- /dev/null
++++ b/src/main/java/com/destroystokyo/paper/meta/PersistentMetaContainer.java
+@@ -0,0 +1,54 @@
++package com.destroystokyo.paper.meta;
++
++import org.bukkit.inventory.ItemStack;
++
++import java.security.InvalidParameterException;
++
++/**
++ * Represents an object that has the capability of storing metadata and
++ * able to persist it between object reloads such as server restarts or chunk reloads.
++ *
++ * Limited to data formats that can be serialized.
++ *
++ * Use the Bukkit {@link org.bukkit.configuration.serialization.ConfigurationSerializable} API to make
++ * your custom data objects safe for persistence.
++ *
++ * This data will not be removed if your plugin is removed. You must handle clean up on your own.
++ * 
++ * This API is not thread safe!
++ */
++public interface PersistentMetaContainer extends MetaContainer {
++
++    @Override
++    default MetaMap getMetaMap(MetaKey key, boolean isWrite) {
++        return getMetaMap((PersistentMetaKey) key, isWrite);
++    }
++
++    PersistentMetaMap getMetaMap(PersistentMetaKey key, boolean isWrite);
++
++    default <T extends PersistentMetaMap> T setMeta(PersistentMetaKey key, PersistentMetaMap value) {
++        if (!Meta.isValidPersistentMeta(value)) {
++            throw new InvalidParameterException();
```

Comment:
Maybe use IllegalStatementException instead of `java.security.InvalidParameterException` here, latter seems to be more used on security/cryptographic/whatever APIs in Java, while IllegalStatementException is for more general purpose code.

Snippet from InvalidParameterException javadoc:
>This exception, designed for use by the JCA/JCE engine classes, is thrown when an invalid parameter is passed to a method.

Reply:
`IllegalArgumentException` makes more sense from a naming standpoint

Reply:
Or that, forgot about it. Thanks

Reply:
Yeah I meant argument. Will change